src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java

```diff
@@ -0,0 +1,2161 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.stream.Collectors;
+
+import org.antlr.v4.runtime.BufferedTokenStream;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.grammar.java.CheckstyleJavaParser;
+import com.puppycrawl.tools.checkstyle.grammar.java.CheckstyleJavaParserBaseVisitor;
+import com.puppycrawl.tools.checkstyle.grammar.java.JavaLexer;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * Visitor class used to build Checkstyle's Java AST.
```

Comment:
Does it make sense to expand this javadoc to explain to future contributors to say how to expand new grammar changes?

I assume there is some general pattern to the madness I see below.

Reply:
I have ordered the visit methods to match the order that parser rules appear in the grammar. I am currently working on a solution for keeping it this way; sadly, `getDeclaredMethods()` does not guarantee the order of the `Method[]` that it returns. I plan to begin implementing a "custom check" of sorts to extract the order of the visit methods from `JavaAstVisitor.java` and compare them to the order of the parser rules. I will have this done by the end of the week.

Reply:
@rnveach if you have an idea of an easy way to extract ordered `visit..()` methods from `JavaAstVisitor.java`, please share.

Reply:
For source order of methods, you would need to write a checkstyle check or something similar.

Reply:
Javadoc expanded.